README.md
# chatgpython

## Instalation

```commandline
pip install chatgpython
```

## Use

```python
from chatgptonic import ChatGPT

chat = ChatGPT("YOUR_API_KEY")

print(chat.send("Some answer"))  # Returns the API return
print(chat.just_chat("Some answer"))  # Returns just the ChatGPT response text
```




CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

* The following tags must be used:
    Added for new features.
    Changed for changes in existing functionality.
    Deprecated for soon-to-be removed features.
    Removed for now removed features.
    Fixed for any bug fixes.
    Security in case of vulnerabilities.

Format message: date_at issue_id message changes
example: * date_at - Task 0002 - fixed some bug.

## [Unreleased]

### Changed

### Added

* 2023-03-14 - Add CI to run tests and style-check

### Fixed

### Removed


tests/conftest.py
import pytest
from chatgptonic.integration import ChatGPT


@pytest.fixture()
def chatgpt():
    return ChatGPT("some_api_key")


tests/test_chatgpt_integration.py
import pytest
from unittest.mock import patch, call

from chatgptonic.integration import ChatGPTMessageException, ChatGPT


class TestChatGPTMessageException:
    def test_parent_class(self):
        assert issubclass(ChatGPTMessageException, Exception)


class TestChatGPT:
    def test_init_without_optional_fields(self, chatgpt):
        assert chatgpt.api_key == "some_api_key"
        assert chatgpt.chat_model == "gpt-3.5-turbo"
        assert chatgpt.role == "user"

    def test_init_with_optional_fields(self):
        chat = ChatGPT("some_api_key", "some_chat_model", "some_role")

        assert chat.api_key == "some_api_key"
        assert chat.chat_model == "some_chat_model"
        assert chat.role == "some_role"

    def test_get_headers(self, chatgpt):
        headers = chatgpt._get_headers()

        assert headers == {
            "Content-Type": "application/json",
            "Authorization": "Bearer some_api_key",
        }

    def test_get_body(self, chatgpt):
        body = chatgpt._get_body("some_message", creativity=0.8)

        assert body == {
            "model": chatgpt.chat_model,
            "messages": [{"role": chatgpt.role, "content": "some_message"}],
            "temperature": 0.8,
        }

    @patch("chatgptonic.integration.json")
    @patch.object(ChatGPT, "_get_body")
    @patch.object(ChatGPT, "_get_headers")
    def test_get_request(self, mock_get_headers, mock_get_body, mock_json, chatgpt):
        request = chatgpt._get_request("some_message", creativity=0.9)
        mock_get_body.assert_called_once_with("some_message", 0.9)
        mock_get_headers.assert_called_once()
        mock_json.dumps.assert_called_once_with(mock_get_body.return_value)

        assert request == {
            "url": chatgpt.chat_url,
            "data": mock_json.dumps.return_value,
            "headers": mock_get_headers.return_value,
        }

    @patch("chatgptonic.integration.json")
    @patch("chatgptonic.integration.requests")
    @patch.object(ChatGPT, "_get_request", return_value={"url": "some_url"})
    def test_send_successfully(
        self, mock_get_request, mock_requests, mock_json, chatgpt
    ):
        mock_requests.post.return_value.status_code = 200
        response = chatgpt.send("some_message", creativity=1)

        mock_get_request.assert_called_once_with("some_message", 1)
        mock_requests.post.assert_called_once_with(url="some_url")

        mock_json.loads.assert_called_once_with(mock_requests.post.return_value.content)

        assert response == mock_json.loads.return_value

    @patch("chatgptonic.integration.json")
    @patch("chatgptonic.integration.requests")
    @patch.object(ChatGPT, "_get_request", return_value={"url": "some_url"})
    def test_send_failure(self, mock_get_request, mock_requests, mock_json, chatgpt):
        mock_requests.post.return_value.status_code = 400
        mock_json.loads.return_value = {"error": {"message": "some_error_message"}}

        with pytest.raises(ChatGPTMessageException) as error:
            chatgpt.send("some_message", creativity=1)

        mock_get_request.assert_called_once_with("some_message", 1)
        mock_requests.post.assert_called_once_with(url="some_url")

        mock_json.loads.assert_called_once_with(mock_requests.post.return_value.content)

        assert str(error.value) == (
            "Something went wrong with this error message: some_error_message"
        )

    @patch.object(
        ChatGPT,
        "send",
        return_value={"choices": [{"message": {"content": "\n\nsome_response"}}]},
    )
    def test_just_chat(self, mock_send, chatgpt):
        response = chatgpt.just_chat("some_message", creativity=0.4)
        mock_send.assert_called_once_with("some_message", creativity=0.4)

        assert response == "some_response"

    @patch.object(ChatGPT, "just_chat", return_value="some_gpt_message")
    @patch("chatgptonic.integration.input")
    @patch("chatgptonic.integration.print")
    def test_interactive_chat_with_quit_input(
        self, _mock_print, mock_input, mock_just_chat, chatgpt
    ):
        mock_input.side_effect = ["some_input", "c"]
        chatgpt.start_interactive_chat()
        mock_input.assert_has_calls(
            [
                call("\033[92m [0.] You: \033[0m"),
                call("\033[92m [1.] You: \033[0m"),
            ]
        )
        mock_just_chat.assert_called_once_with("some_input")


chatgptonic/integration.py
import json

import requests


class ChatGPTMessageException(Exception):
    pass


class ChatGPT:
    def __init__(
        self, api_key: str, chat_model: str = "gpt-3.5-turbo", role: str = "user"
    ):
        self.api_key: str = api_key

        self.chat_model: str = chat_model
        self.role: str = role
        self.chat_url: str = "https://api.openai.com/v1/chat/completions"

    def _get_headers(self) -> dict:
        return {
            "Content-Type": "application/json",
            "Authorization": f"Bearer {self.api_key}",
        }

    def _get_body(self, message: str, creativity: float) -> dict:
        return {
            "model": self.chat_model,
            "messages": [{"role": self.role, "content": message}],
            "temperature": creativity,
        }

    def _get_request(self, message: str, creativity: float):
        body = self._get_body(message, creativity)
        headers = self._get_headers()

        return {"url": self.chat_url, "data": json.dumps(body), "headers": headers}

    def send(self, message: str, creativity: float = 0.7) -> dict:
        request = self._get_request(message, creativity)

        response = requests.post(**request)
        response_json = json.loads(response.content)

        if response.status_code == 200:
            return response_json
        else:
            errors = response_json["error"]["message"]

        raise ChatGPTMessageException(
            f"Something went wrong with this error message: {errors}"
        )

    def just_chat(self, message: str, creativity: float = 0.7) -> str:
        response = self.send(message, creativity=creativity)

        return response["choices"][0]["message"]["content"].replace("\n", "")

    def start_interactive_chat(self):
        count = 0
        while True:
            message = input(f"\033[92m [{count}.] You: \033[0m")
            if message == "c":
                break
            print(f"\033[93m [{count}.] Thinking...\033[0m")
            response = self.just_chat(message)
            print(f"\033[94m [{count}.] ChatGPT: {response}\033[0m")
            print(f"\033[96m [{count}.] ---------------------------\033[0m")
            count += 1
